File: scripts/addUsersToFirebase.js

const admin = require('firebase-admin');
const serviceAccount = require('./serviceAccountKey.json'); // Make sure this file exists in the same directory

// Initialize Firebase Admin SDK
try {
    admin.initializeApp({
        credential: admin.credential.cert(serviceAccount)
    });
    console.log("Firebase Admin SDK initialized successfully.");
} catch (error) {
    console.error("Error initializing Firebase Admin SDK:", error);
    console.error("Please ensure 'serviceAccountKey.json' is in the 'scripts' directory and is valid.");
    process.exit(1);
}

const db = admin.firestore();
const auth = admin.auth();

const usersToAddRaw = [
    // employeeId, firstNameEn, lastNameEn, fullNameTh, role
    ["100024", "Issara", "Ponjarone", "อิสระ พรเจริญ", "BimCoordinate"],
    ["100097", "Rangsan", "Manosan", "รังสรรค์ มาโนสาร", "BimManager"],
    ["100105", "Watcharapong", "Wongta", "วัชรพงษ์ วงศ์ตา", "BimCoordinate"],
    ["100128", "Thanaphat", "Namwong", "ธนภัทร น้ำวงศ์", "BimLeader"],
    ["100136", "Ekasith", "Anburi", "เอกสิทธิ์ อั้นบุรี", "BimLeader"],
    ["100454", "Siwaporn", "Pankhui", "ศิวพร พานขุย", "BimCoordinate"],
    ["100646", "Sithichocke", "Nguanchoo", "สิทธิโชค เงื่อนชู", "BimLeader"],
    ["100769", "Surasak", "Singthongla", "สุรศักดิ์ สิงห์ทองหล้า", "BimCoordinate"],
    ["100884", "Jutatip", "Srirat", "จุฑาทิพย์ ศรีรัตน์", "BimCoordinate"],
    ["101020", "Pattanan", "Junlamunee", "พัฒนันท์ จันทร์ละมุนี", "BimCoordinate"],
    ["101477", "Worawut", "Nawongsa", "วรวุฒิ นาวงษา", "BimModeler"],
    ["101486", "Onnicha", "Kamsee", "อรนิชา คำสี", "BimModeler"],
    ["101508", "Bussaya", "Thapanya", "บุษยา ทาปัญญา", "BimModeler"],
    ["101760", "Narongsak", "Buttumpan", "ณรงค์ศักดิ์ บุญธัมปานนท์", "BimModeler"],
    ["101783", "Pattarawichaya", "Chusripetch", "ภัทรวิชยา ชูศรีเพชร", "BimModeler"],
    ["101794", "Pasakorn", "Phopumnak", "ปาสกร ภพพำนัก", "BimModeler"],
    ["101795", "Sutida", "Srijampa", "สุธิดา ศรีจำปา", "BimModeler"],
    ["101782", "nattagun", "talabnak", "ณัฐกุล ตลับนาค", "BimModeler"],
    ["101780", "Thitikorn", "Bumrungkate", "ฐิติกานต์ บำรุงเกตุ", "BimModeler"],
];

function generateUserData(employeeId, firstNameEn, lastNameEn, fullNameTh, role) {
    const fullNameEn = `${firstNameEn} ${lastNameEn}`;

    // Generate username: first name in lowercase + '.' + first letter of last name in lowercase
    const username = `${firstNameEn.toLowerCase()}.${lastNameEn.toLowerCase()[0]}`;

    // Password is the employeeId
    const password = employeeId;

    return {
        employeeId: employeeId,
        fullNameTh: fullNameTh,
        fullNameEn: fullNameEn,
        user: username,
        password: password, // Storing password in Firestore for simplicity, but for production, avoid storing plaintext passwords.
        role: role // Add the role here
    };
}

async function addUserToFirestoreAndAuth(userData) {
    const employeeId = userData.employeeId;
    const usernameEmail = `${userData.user}@yourproject.com`; // Use a dummy domain for username to act as email for Auth
    const password = userData.password;
    const role = userData.role;

    console.log(`Processing user: ${userData.fullNameEn} (ID: ${employeeId}) with role: ${role}`);

    // 1. Add/Update user in Firestore
    try {
        const docRef = db.collection('Users').doc(employeeId);
        await docRef.set(userData);
        console.log(`Firestore: User '${userData.fullNameEn}' (ID: ${employeeId}) added/updated.`);
    } catch (e) {
        console.error(`Firestore Error for ${userData.fullNameEn}:`, e);
    }

    // 2. Create/Update user in Firebase Authentication and set custom claims
    try {
        let userRecord;
        try {
            userRecord = await auth.getUserByEmail(usernameEmail);
            console.log(`Auth: User '${usernameEmail}' already exists. Updating custom claims.`);
        } catch (error) {
            if (error.code === 'auth/user-not-found') {
                userRecord = await auth.createUser({
                    email: usernameEmail,
                    password: password,
                    displayName: userData.fullNameEn,
                    uid: employeeId // Use employeeId as UID for easy linking
                });
                console.log(`Auth: User '${usernameEmail}' created successfully with UID: ${userRecord.uid}`);
            } else {
                throw error; // Re-throw other authentication errors
            }
        }
        await auth.setCustomUserClaims(userRecord.uid, { role: role });
        console.log(`Auth: Custom claim 'role' set to '${role}' for user ${usernameEmail}.`);
    } catch (e) {
        console.error(`Auth Error for user ${usernameEmail}:`, e);
    }
}

async function main() {
    console.log("Starting user data population...");
    for (const [employeeId, firstNameEn, lastNameEn, fullNameTh, role] of usersToAddRaw) {
        const userData = generateUserData(employeeId, firstNameEn, lastNameEn, fullNameTh, role);
        await addUserToFirestoreAndAuth(userData);
    }
    console.log("User data population complete.");
    console.log("\nIMPORTANT: For Firebase Authentication, the 'user' field is used as an email with a dummy domain '@yourproject.com'.");
    console.log("You will need to ensure your frontend login logic uses this format, or adjust the script to use a real domain if you have one.");
    console.log("Also, for production applications, do NOT store plaintext passwords in Firestore as done in this example. Use Firebase Authentication for secure password management.");
    console.log("To access custom claims (like role) in your frontend, you'll need to re-authenticate the user or force a token refresh after claims are set.");
    console.log("For example, after login, you might call 'firebase.auth().currentUser.getIdToken(true)' to get the latest token with claims.");
}

main().catch(console.error);

File: scripts/add_users_to_firebase.py

import firebase_admin
from firebase_admin import credentials, firestore, auth
import os

# Path to your service account key file
# Make sure this file is in the same directory as this script, or provide the full path
SERVICE_ACCOUNT_KEY_PATH = os.path.join(os.path.dirname(__file__), 'serviceAccountKey.json')

# Initialize Firebase Admin SDK
try:
    cred = credentials.Certificate(SERVICE_ACCOUNT_KEY_PATH)
    firebase_admin.initialize_app(cred)
    db = firestore.client()
    print("Firebase Admin SDK initialized successfully.")
except Exception as e:
    print(f"Error initializing Firebase Admin SDK: {e}")
    print("Please ensure 'serviceAccountKey.json' is in the 'scripts' directory and is valid.")
    exit()

users_to_add_raw = [
    ("100024", "Issara", "Ponjarone", "BimCoordinate"),
    ("100097", "Rangsan", "Manosan", "BimManager"),
    ("100105", "Watcharapong", "Wongta", "BimCoordinate"),
    ("100128", "Thanaphat", "Namwong", "BimLeader"),
    ("100136", "Ekasith", "Anburi", "BimLeader"),
    ("100454", "Siwaporn", "Pankhui", "BimCoordinate"),
    ("100646", "Sithichocke", "Nguanchoo", "BimLeader"),
    ("100769", "Surasak", "Singthongla", "BimCoordinate"),
    ("100884", "Jutatip", "Srirat", "BimCoordinate"),
    ("101020", "Pattanan", "Junlamunee", "BimCoordinate"),
    ("101477", "Worawut", "Nawongsa", "BimModeler"),
    ("101486", "Onnicha", "Kamsee", "BimModeler"),
    ("101508", "Bussaya", "Thapanya", "BimModeler"),
    ("101760", "Narongsak", "Buttumpan", "BimModeler"),
    ("101783", "Pattarawichaya", "Chusripetch", "BimModeler"),
    ("101794", "Pasakorn", "Phopumnak", "BimModeler"),
    ("101795", "Sutida", "Srijampa", "BimModeler"),
    ("101782", "nattagun", "talabnak", "BimModeler"),
    ("101780", "Thitikorn", "Bumrungkate", "BimModeler"),
]

def generate_user_data(employeeId, firstNameEn, lastNameEn, role):
    fullNameEn = f"{firstNameEn} {lastNameEn}"
    fullNameTh = f"{firstNameEn} {lastNameEn}" # Assuming Thai name is same as English name if not provided
    
    # Generate username: first name in lowercase + '.' + first letter of last name in lowercase
    username = f"{firstNameEn.lower()}.{lastNameEn.lower()[0]}"
    
    # Password is the employeeId
    password = employeeId

    return {
        "employeeId": employeeId,
        "fullNameTh": fullNameTh,
        "fullNameEn": fullNameEn,
        "user": username,
        "password": password, # Storing password in Firestore for simplicity, but for production, avoid storing plaintext passwords.
        "role": role # Add the role here
    }

def add_user_to_firestore_and_auth(user_data):
    employee_id = user_data["employeeId"]
    username_email = f"{user_data['user']}@yourproject.com" # Use a dummy domain for username to act as email for Auth
    password = user_data["password"]
    role = user_data["role"]

    print(f"Processing user: {user_data['fullNameEn']} (ID: {employee_id}) with role: {role}")

    # 1. Add/Update user in Firestore
    try:
        doc_ref = db.collection('Users').document(employee_id)
        doc_ref.set(user_data)
        print(f"Firestore: User '{user_data['fullNameEn']}' (ID: {employee_id}) added/updated.")
    except Exception as e:
        print(f"Firestore Error for {user_data['fullNameEn']}: {e}")

    # 2. Create user in Firebase Authentication and set custom claims
    try:
        user_record = auth.get_user_by_email(username_email)
        print(f"Auth: User '{username_email}' already exists. Updating custom claims.")
        auth.set_custom_user_claims(user_record.uid, {'role': role})
        print(f"Auth: Custom claim 'role' set to '{role}' for user {username_email}.")
    except auth.UserNotFoundError:
        try:
            user_record = auth.create_user(
                email=username_email,
                password=password,
                display_name=user_data["fullNameEn"],
                uid=employee_id # Use employeeId as UID for easy linking
            )
            print(f"Auth: User '{username_email}' created successfully with UID: {user_record.uid}")
            auth.set_custom_user_claims(user_record.uid, {'role': role})
            print(f"Auth: Custom claim 'role' set to '{role}' for new user {username_email}.")
        except Exception as e:
            print(f"Auth Error creating user {username_email}: {e}")
    except Exception as e:
        print(f"Auth Error checking user {username_email}: {e}")


if __name__ == "__main__":
    print("Starting user data population...")
    for emp_id, first_name, last_name, role in users_to_add_raw:
        user_data = generate_user_data(emp_id, first_name, last_name, role)
        add_user_to_firestore_and_auth(user_data)
    print("User data population complete.")
    print("\nIMPORTANT: For Firebase Authentication, the 'user' field is used as an email with a dummy domain '@yourproject.com'.")
    print("You will need to ensure your frontend login logic uses this format, or adjust the script to use a real domain if you have one.")
    print("Also, for production applications, do NOT store plaintext passwords in Firestore as done in this example. Use Firebase Authentication for secure password management.")
    print("To access custom claims (like role) in your frontend, you'll need to re-authenticate the user or force a token refresh after claims are set.")
    print("For example, after login, you might call 'firebase.auth().currentUser.getIdToken(true)' to get the latest token with claims.")
